File: EHO.m
clear all;
clc;

tic;

%% Data
%data001 = get(handles.uitable1,'UserData');
%data=data001{1};
%NamaBarang=data001{2};
%Kapasitas=str2num(get(handles.edit6,'string'));
%Modal=str2num(get(handles.edit7,'string'));


maxpeso=165;
peso=[ 23 31 29 44 53 38 63 85 89 82];
custo=[92 57 49 68 60 43 67 84 87 72];
%otimo=[1 1 1 1 0 1 0 0 0 0];
% 0.064184 seconds
dim=size(peso,2)

%% Parameter
nClan= 1; %Número de clãs (ou grupos de soluções).
nci=10; %Número de indivíduos em cada clã.
alpha=0.5; %Parâmetros de ajuste usados na atualização das posições das soluções.
beta=0.5;  %Parâmetros de ajuste usados na atualização das posições das soluções.
MaxGen=30; %Número máximo de gerações.

%% Inisialisasi
X=rand(nClan*nci,dim); % Gera uma matriz de números aleatórios contínuos entre 0 e 1
Y=round(X); %Arredonda os valores para 0 ou 1


%% Inisialisasi
for i=1:nClan %itera cada clan i
    for j=1:nci %itera cada elefante j do clan i
        Berat(i,j)=sum(Y((i-1)*nci+j,:).*peso); %calcua o peso de cada solução
        Beli(i,j)=sum(Y((i-1)*nci+j,:).*custo); %calcula o custo de cada solução
        while Berat(i,j)>maxpeso
            pss=find(Y((i-1)*nci+j,:)==1);
            r1=ceil(rand*length(pss));
            X((i-1)*nci+j,pss(r1))=X((i-1)*nci+j,pss(r1))/2;
            Y((i-1)*nci+j,:)=round(X((i-1)*nci+j,:));
            Berat(i,j)=sum(Y((i-1)*nci+j,:).*peso);
            Beli(i,j)=sum(Y((i-1)*nci+j,:).*custo);
        end
        Fitness(i,j)=Beli(i,j);
    end
    best=find(Fitness(i,:)==max(Fitness(i,:)));
    Matriarch(i,:)=X((i-1)*nci+best(1),:);
    FitMat(i)=max(Fitness(i,:));
end

%%
bsf(1)=max(FitMat);
inon=0;
t=0;

while t<MaxGen
    %Update posisi
    for i=1:nClan
        Xctr=sum(X((i-1)*nci+(1:nci),:))/nci;
        X((i-1)*nci+(1:nci),:)=X((i-1)*nci+(1:nci),:)+alpha*(repmat(Matriarch(i,:),nci,1)-X((i-1)*nci+(1:nci),:)).*rand(nci,dim);
        for j=1:nci
            if Y((i-1)*nci+j,:)==round(Matriarch(i,:))
                X((i-1)*nci+j,:)=X((i-1)*nci+j,:)+beta*Xctr.*rand(1,dim);
            end
            if min(X((i-1)*nci+j,:))<0
                X((i-1)*nci+j,:)=(X((i-1)*nci+j,:)-min(X((i-1)*nci+j,:)))/(max(X((i-1)*nci+j,:))-min(X((i-1)*nci+j,:)));
            else
                X((i-1)*nci+j,:)=X((i-1)*nci+j,:)/max(X((i-1)*nci+j,:));
            end
            Y((i-1)*nci+j,:)=round(X((i-1)*nci+j,:));
            Berat(i,j)=sum(Y((i-1)*nci+j,:).*peso);
            Beli(i,j)=sum(Y((i-1)*nci+j,:).*custo);
            while Berat(i,j)>maxpeso
                pss=find(Y((i-1)*nci+j,:)==1);
                r1=ceil(rand*length(pss));
                X((i-1)*nci+j,pss(r1))=X((i-1)*nci+j,pss(r1))/2;
                Y((i-1)*nci+j,:)=round(X((i-1)*nci+j,:));
                Berat(i,j)=sum(Y((i-1)*nci+j,:).*peso);
                Beli(i,j)=sum(Y((i-1)*nci+j,:).*custo);
            end
            Fitness(i,j)=Beli(i,j);
        end
        %Replace worst
        worst=find(Fitness(i,:)==min(Fitness(i,:)));
        if length(worst)==nci
            worst=ceil(0.75*nci):nci;
        end
        X((i-1)*nci+worst,:)=repmat(min(X((i-1)*nci+(1:nci),:)),length(worst),1)+...
            repmat((max(X((i-1)*nci+(1:nci),:))-min(X((i-1)*nci+(1:nci),:)))+1,length(worst),1).*rand(length(worst),dim);
        for k=1:length(worst)
            if min(X((i-1)*nci+worst(k),:))<0
                X((i-1)*nci+worst(k),:)=(X((i-1)*nci+worst(k),:)-min(X((i-1)*nci+worst(k),:)))/(max(X((i-1)*nci+worst(k),:))-min(X((i-1)*nci+worst(k),:)));
            else
                X((i-1)*nci+worst(k),:)=X((i-1)*nci+worst(k),:)/max(X((i-1)*nci+worst(k),:));
            end
            Y((i-1)*nci+worst(k),:)=round(X((i-1)*nci+worst(k),:));
            Berat(i,worst(k))=sum(Y((i-1)*nci+worst(k),:).*peso);
            Beli(i,worst(k))=sum(Y((i-1)*nci+worst(k),:).*custo);
            while Berat(i,worst(k))>maxpeso
                pss=find(Y((i-1)*nci+worst(k),:)==1);
                r1=ceil(rand*length(pss));
                X((i-1)*nci+worst(k),pss(r1))=X((i-1)*nci+worst(k),pss(r1))/2;
                Y((i-1)*nci+worst(k),:)=round(X((i-1)*nci+worst(k),:));
                Berat(i,worst(k))=sum(Y((i-1)*nci+worst(k),:).*peso);
                Beli(i,worst(k))=sum(Y((i-1)*nci+worst(k),:).*custo);
            end
            Fitness(i,worst(k))= Beli(i,worst(k));
        end
        %Update Matriarch
        if max(Fitness(i,:))>FitMat(i)
            best=find(Fitness(i,:)==max(Fitness(i,:)));
            Matriarch(i,:)=X((i-1)*nci+best(1),:);
            FitMat(i)=max(Fitness(i,:));
        end
    end
    %Plot
    bsf(t+2)=max(FitMat);
    if bsf(t+2)~=bsf(t+1)
        inon=t+1;
    end
    plot(0:length(bsf)-1,bsf,'r','LineWidth',2);
    line(inon,bsf(inon+1),'Marker','s','MarkerEdgeColor','k','MarkerFaceColor','b','MarkerSize',6);
    if inon<t*0.75
        text(inon,bsf(inon+1),sprintf(['\n\n Iter: ' num2str(inon) '\n Fit: ' sprintf('%12d',bsf(inon+1)) ',-']));
    else
        text(inon,bsf(inon+1),sprintf([' Iter: ' num2str(inon) '\nFit: ' sprintf('%12d',bsf(inon+1)) ',-''\n\n']),'HorizontalAlignment','Right');
    end
    ylim([min(bsf) max(bsf)*1.08]);
    %set(handles.axes1,'XColor',[1 1 1],'YColor',[1 11],'FontSize',8,'FontWeight','bold');
    xlabel('Iterasi'); ylabel('Total Profit (Fitness)');
    pause(0.00001);
    %Next iter
    t=t+1;
end
best=find(FitMat==max(FitMat))
FinalSol=round(Matriarch(best(1),:))
toc;

File: EHO.py
import numpy as np

# Dados do problema
max_peso = 6404180
peso = np.array([382745, 799601, 909247, 729069, 467902, 44328, 34610, 698150, 823460, 903959, 
                 853665, 551830, 610856, 670702, 488960, 951111, 323046, 446298, 931161, 31385, 
                 496951, 264724, 224916, 169684])
custo = np.array([825594, 1677009, 1676628, 1523970, 943972, 97426, 69666, 1296457, 1679693, 1902996,
                  1844992, 1049289, 1252836, 1319836, 953277, 2067538, 675367, 853655, 1826027, 65731,
                  901489, 577243, 466257, 369261])

# Parâmetros
n_cla = 1  # Número de clãs (grupos de soluções)
n_individuos = 1000  # Número de indivíduos em cada clã
alpha = 0.5  # Parâmetro de ajuste para a atualização das posições
beta = 0.5   # Parâmetro de ajuste para a atualização das posições
max_geracoes = 300  # Número máximo de gerações

# Inicialização
dim = len(peso)  # Dimensão do problema (número de itens)

# Geração inicial aleatória de soluções (números entre 0 e 1)
X = np.random.rand(n_cla * n_individuos, dim)
# Arredondamento para 0 ou 1 (soluções binárias)
Y = np.round(X)

# Inicialização das soluções
peso_solucao = np.zeros((n_cla, n_individuos))
valor_solucao = np.zeros((n_cla, n_individuos))
aptidao = np.zeros((n_cla, n_individuos))
matriarca = np.zeros((n_cla, dim))
fitness_matriarca = np.zeros(n_cla)

for i in range(n_cla):  # Itera sobre cada clã
    for j in range(n_individuos):  # Itera sobre cada elefante (solução) do clã
        # Calcula o peso total da solução
        peso_solucao[i, j] = np.sum(Y[i * n_individuos + j, :] * peso)
        # Calcula o custo total da solução
        valor_solucao[i, j] = np.sum(Y[i * n_individuos + j, :] * custo)
        
        # Se o peso excede o limite, ajusta a solução
        while peso_solucao[i, j] > max_peso:
            itens_selecionados = np.where(Y[i * n_individuos + j, :] == 1)[0]  # Itens na solução
            item_aleatorio = np.random.choice(itens_selecionados)  # Seleciona item aleatório
            # Remove o item escolhido
            X[i * n_individuos + j, item_aleatorio] /= 2
            Y[i * n_individuos + j, :] = np.round(X[i * n_individuos + j, :])  # Arredonda
            # Recalcula o peso e o valor da solução ajustada
            peso_solucao[i, j] = np.sum(Y[i * n_individuos + j, :] * peso)
            valor_solucao[i, j] = np.sum(Y[i * n_individuos + j, :] * custo)
        
        # Fitness é igual ao valor total da solução
        aptidao[i, j] = valor_solucao[i, j]
    
    # A melhor solução do clã (matriarca) é a de maior fitness
    melhor = np.argmax(aptidao[i, :])  # Índice da solução de maior fitness
    matriarca[i, :] = X[i * n_individuos + melhor, :]  # Atualiza matriarca
    fitness_matriarca[i] = aptidao[i, melhor]  # Fitness da matriarca
    
    # Otimização
melhor_geral = [np.max(fitness_matriarca)]  # Armazena a melhor solução
t = 0  # Contador de gerações

while t < max_geracoes:
    # Atualiza as posições dos elefantes em cada clã
    for i in range(n_cla):
        # Calcula o centro do clã (média das posições)
        centro_cla = np.sum(X[i * n_individuos:(i + 1) * n_individuos, :], axis=0) / n_individuos
        
        # Atualiza as posições dos elefantes em direção à matriarca
        X[i * n_individuos:(i + 1) * n_individuos, :] += alpha * (np.tile(matriarca[i, :], (n_individuos, 1)) - X[i * n_individuos:(i + 1) * n_individuos, :]) * np.random.rand(n_individuos, dim)
        
        # Ajusta a posição dos elefantes
        for j in range(n_individuos):
            if np.array_equal(Y[i * n_individuos + j, :], np.round(matriarca[i, :])):
                X[i * n_individuos + j, :] += beta * centro_cla * np.random.rand(dim)
                
            # Normaliza a posição dos elefantes
            if np.min(X[i * n_individuos + j, :]) < 0:
                X[i * n_individuos + j, :] = (X[i * n_individuos + j, :] - np.min(X[i * n_individuos + j, :])) / (np.max(X[i * n_individuos + j, :]) - np.min(X[i * n_individuos + j, :]))
            else:
                X[i * n_individuos + j, :] = X[i * n_individuos + j, :] / np.max(X[i * n_individuos + j, :])
            Y[i * n_individuos + j, :] = np.round(X[i * n_individuos + j, :])
            
            # Recalcula o peso e o valor das soluções
            peso_solucao[i, j] = np.sum(Y[i * n_individuos + j, :] * peso)
            valor_solucao[i, j] = np.sum(Y[i * n_individuos + j, :] * custo)
            
            # Se o peso excede o limite, ajusta a solução
            while peso_solucao[i, j] > max_peso:
                itens_selecionados = np.where(Y[i * n_individuos + j, :] == 1)[0]
                item_aleatorio = np.random.choice(itens_selecionados)
                X[i * n_individuos + j, item_aleatorio] /= 2
                Y[i * n_individuos + j, :] = np.round(X[i * n_individuos + j, :])
                peso_solucao[i, j] = np.sum(Y[i * n_individuos + j, :] * peso)
                valor_solucao[i, j] = np.sum(Y[i * n_individuos + j, :] * custo)
            
            aptidao[i, j] = valor_solucao[i, j]

        # Substitui as piores soluções do clã
        pior = np.where(aptidao[i, :] == np.min(aptidao[i, :]))[0]
        if len(pior) == n_individuos:
            pior = np.arange(int(0.75 * n_individuos), n_individuos)
        X[i * n_individuos + pior, :] = np.tile(np.min(X[i * n_individuos:(i + 1) * n_individuos, :], axis=0), (len(pior), 1)) + (np.tile(np.max(X[i * n_individuos:(i + 1) * n_individuos, :], axis=0) - np.min(X[i * n_individuos:(i + 1) * n_individuos, :], axis=0) + 1, (len(pior), 1))) * np.random.rand(len(pior), dim)

        # Ajusta as novas soluções geradas
        for k in range(len(pior)):
            if np.min(X[i * n_individuos + pior[k], :]) < 0:
                X[i * n_individuos + pior[k], :] = (X[i * n_individuos + pior[k], :] - np.min(X[i * n_individuos + pior[k], :])) / (np.max(X[i * n_individuos + pior[k], :]) - np.min(X[i * n_individuos + pior[k], :]))
            else:
                X[i * n_individuos + pior[k], :] = X[i * n_individuos + pior[k], :] / np.max(X[i * n_individuos + pior[k], :])
            Y[i * n_individuos + pior[k], :] = np.round(X[i * n_individuos + pior[k], :])

            # Recalcula o peso e o valor das soluções
            peso_solucao[i, pior[k]] = np.sum(Y[i * n_individuos + pior[k], :] * peso)
            valor_solucao[i, pior[k]] = np.sum(Y[i * n_individuos + pior[k], :] * custo)

            # Se o peso excede o limite, ajusta a solução
            while peso_solucao[i, pior[k]] > max_peso:
                itens_selecionados = np.where(Y[i * n_individuos + pior[k], :] == 1)[0]
                item_aleatorio = np.random.choice(itens_selecionados)
                X[i * n_individuos + pior[k], item_aleatorio] /= 2
                Y[i * n_individuos + pior[k], :] = np.round(X[i * n_individuos + pior[k], :])
                peso_solucao[i, pior[k]] = np.sum(Y[i * n_individuos + pior[k], :] * peso)
                valor_solucao[i, pior[k]] = np.sum(Y[i * n_individuos + pior[k], :] * custo)
            
            aptidao[i, pior[k]] = valor_solucao[i, pior[k]]

        # Atualiza a matriarca se houver uma nova melhor solução no clã
        if np.max(aptidao[i, :]) > fitness_matriarca[i]:
            melhor = np.argmax(aptidao[i, :])
            matriarca[i, :] = X[i * n_individuos + melhor, :]
            fitness_matriarca[i] = np.max(aptidao[i, :])

    # Armazena a melhor solução da geração
    melhor_geral.append(np.max(fitness_matriarca))

    # Próxima geração
    t += 1

# Exibe a melhor solução final
melhor = np.argmax(fitness_matriarca)
solucao_final = np.round(matriarca[melhor, :])  # A solução final encontrada

print("Solução final:", solucao_final)

valor_total_solucao_final = np.sum(solucao_final * custo)

print("Valor total da solução final:", valor_total_solucao_final)
